lib/httputils/statconn.go

```diff
@@ -27,47 +31,49 @@ type connStateMetrics struct {
 	connBytesWritten *metrics.Counter
 }
 
-func (cms *connStateMetrics) init(metricPrefix string) {
-	cms.dialsTotal = metrics.NewCounter(fmt.Sprintf(`%s_dials_total`, metricPrefix))
-	cms.dialErrors = metrics.NewCounter(fmt.Sprintf(`%s_dial_errors_total`, metricPrefix))
-	cms.conns = metrics.NewCounter(fmt.Sprintf(`%s_conns`, metricPrefix))
-
-	cms.connReadsTotal = metrics.NewCounter(fmt.Sprintf(`%s_conn_reads_total`, metricPrefix))
-	cms.connWritesTotal = metrics.NewCounter(fmt.Sprintf(`%s_conn_writes_total`, metricPrefix))
-	cms.connReadErrors = metrics.NewCounter(fmt.Sprintf(`%s_conn_read_errors_total`, metricPrefix))
-	cms.connWriteErrors = metrics.NewCounter(fmt.Sprintf(`%s_conn_write_errors_total`, metricPrefix))
-	cms.connBytesRead = metrics.NewCounter(fmt.Sprintf(`%s_conn_bytes_read_total`, metricPrefix))
-	cms.connBytesWritten = metrics.NewCounter(fmt.Sprintf(`%s_conn_bytes_written_total`, metricPrefix))
+func newStatConnMetrics(metricPrefix string) *statConnMetrics {
+	scm := &statConnMetrics{}
+
+	scm.dialsTotal = metrics.NewCounter(fmt.Sprintf(`%s_dials_total`, metricPrefix))
+	scm.dialErrors = metrics.NewCounter(fmt.Sprintf(`%s_dial_errors_total`, metricPrefix))
+	scm.conns = metrics.NewCounter(fmt.Sprintf(`%s_conns`, metricPrefix))
+
+	scm.connReadsTotal = metrics.NewCounter(fmt.Sprintf(`%s_conn_reads_total`, metricPrefix))
+	scm.connWritesTotal = metrics.NewCounter(fmt.Sprintf(`%s_conn_writes_total`, metricPrefix))
+	scm.connReadErrors = metrics.NewCounter(fmt.Sprintf(`%s_conn_read_errors_total`, metricPrefix))
+	scm.connWriteErrors = metrics.NewCounter(fmt.Sprintf(`%s_conn_write_errors_total`, metricPrefix))
+	scm.connBytesRead = metrics.NewCounter(fmt.Sprintf(`%s_conn_bytes_read_total`, metricPrefix))
+	scm.connBytesWritten = metrics.NewCounter(fmt.Sprintf(`%s_conn_bytes_written_total`, metricPrefix))
+
+	return scm
 }
 
 // GetStatDialFunc returns dial function that supports DNS SRV records,
 // and register stats metrics for conns.
+// GetStatDialFunc should be executed only once for each unique metricPrefix.
```

Comment:
That's not true, GetStatDialFunc() will be called multiple times in some cases, like [creating scraper for each target ](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/01eba1116b52714830257f411eac99fcb550b66d/lib/promscrape/client.go#L75) and [remote write to multiple endpoints](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/01eba1116b52714830257f411eac99fcb550b66d/app/vmagent/remotewrite/client.go#L123) in vmagent.

Reply:
Ok, then my change in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/4d9d6d7d9ff2e741aa58b5e7f370c97d69fb6079 is irrelevant

Reply:
See another update https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6299/commits/699e4ef77c1762715e24a3ec9417b6c2ab58f697

Reply:
Sorry, didn't notice your update before force push, please check https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6299/commits/7b2e73ae4f0eec107434934216f07ab76d580256